### PR TITLE
Issue #2406: Clipboard Functionality [backup]

### LINKED ISSUE #2406: Make it easy for analysts to copy domain request data to clipboard
### Issue description
Right now, Analysts manually copy over domain request information to a Microsoft Planner as part of the review process. This means tediously finding each item listed in the additional context, highlighting the value to copy to clipboard, and then pasting in a template. The template they use is exactly what is in the additional context. This is tedious work that doesn't add value. 

To make analyst lives easier, we want to copy all the relevant data from the domain request into the clipboard matching the format they use. This includes the bolded subtitle names and having everything inline as much as possible to create a condensed view.

### Acceptance criteria

- [ ] There is a "Copy request summary" button added to domain requests in django admin. Button should appear under the "History" button in upper right of page.
- [ ] The button adds the content as described in the below ACs to clipboard
- [ ] Content matches the template in the additional context, the values in <> are meant to provide direction the data that should go in each field
- [ ] One blank line should be added between each head (just like below) and headers should be bold

### Additional context

**Recommendation:** < leave empty >

**Organization Type:** <organization_type>

**Requested Domain:** <requested_domain>

**Current Websites:** 

* <current_website>
* Add Whois links and sources used to confirm that it's an official website for the requesting organization

**Rationale:** < leave empty >

**Alternative Domains:**

* < alternate domain names in a comma separated list>

**Submitter:** 

* <name, title  and contact info>
* Add additional details and links used for verification

**Senior Official:**

* <name, title and contact info in one line>
* Add additional details and links used for verification

**Other Employees:** 

* <contact 1 name, title, contact info>
  * Add additional details and links used for verification
* <contact 2 name, title, contact info>
  * Add additional details and links used for verification 

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by therealslimhsiehdy**
## Ticket 2406

Resolves #2406

## Changes

- change_form_object_tools.html (added a button that copies summary to clipboard if a url is not provided)
- copy-summary.js (added function to copy summary to clipboard)
- contact_detail_list.html (added ids for safer DOM traversing)

## Code Review Verification Steps

code is deployed to [sandbox-rh](https://getgov-rh.app.cloud.gov/admin/registrar/domainrequest/163/change/?_changelist_filters=status__in%3Dsubmitted%252Cin%2Breview%252Caction%2Bneeded#)

- Navigate to /admin
- Open a domain request
- Verify that...
   - [ ] a button appears in the upper right that says "Copy to clipboard".  
   - [ ] Clicking it copies a summary of the Domain Request to clipboard in a format that matches the ticket ACs

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/assets/js/copy-summary.js
Comment (Katherine-Osos) [inline src/registrar/assets/js/copy-summary.js:1]: @CocoByte I don't think the button is copying properly. When I copied details for this request on backup (he-hundred.gov), and pasted into a document, it looked like this: 

Recommendation:
Organization Type: federal
Requested Domain: 163
Existing website(s): 
Rationale:Alternate Domain(s): 
Submitter: 566
Senior Official: 565
Additional Contact(s): 

- For some fields, it's displaying numbers rather than the actual text.
- "Additional contacts," "existing websites," and "alternate domains" appear blank even though there's info for those fields in that request. 
- "Alternate domains" should also  appear on its own line.

Comment (CocoByte) [reply src/registrar/assets/js/copy-summary.js:1]: should be fixed now!

**Comment by dave-kennedy-ecs** [comment]
After I click the copy button and get a confirmation in Chrome that it has been copied to the clipboard, I am not able to Paste into Notepad.  I am testing on backup.

Also, I noticed on backup that the styling of images on the finish-user-profile-setup is different.  I am not sure if something with the backup sandbox, or with this PR
image

**Review by therealslimhsiehdy — COMMENTED** [review]
Looks like it's running into some issues still, but super close!

I created [this domain on backup](https://getgov-backup.app.cloud.gov/admin/registrar/domainrequest/246/change/?_changelist_filters=status__in%3Dsubmitted%252Cin%2Breview%252Caction%2Bneeded#) and am the submitter + have other contacts, but when I copy on the clipboard I get:

**Recommendation:**
**Organization Type:** Federal
**Requested Domain:** govdomainiwant.gov
**Existing website(s):** http://currentwebsite.com, http://currentwebsite01.com
**Rationale:**
**Alternate Domain(s):** altdomain1.gov, altdomain2.gov, altdomain3.gov
**Submitter:** 
**Senior Official:** SeniorFirst SeniorLast, SeniorTitle, Senioremail@so.com
**Additional Contact(s):**

It seems to not be displaying the Submitter or Additional Contact(s)

File: src/registrar/templates/admin/change_form.html
Comment (therealslimhsiehdy) [inline src/registrar/templates/admin/change_form.html:2]: **Nitpick (optional):** Can probs remove the comment here!

**Comment by Katherine-Osos** [comment]
@CocoByte The "additional contacts" are not being copied. Can you look into that? 

Also, instead of a popup confirmation message, can you mimic the established design pattern that's currently used for the  "Copy" buttons next to email addresses?. I.e. button is temporarily replaced by message that reads:

> < checkmark icon > Copied to clipboard

**Comment by CocoByte** [comment]
> @CocoByte The "additional contacts" are not being copied. Can you look into that?
> 
> Also, instead of a popup confirmation message, can you mimic the established design pattern that's currently used for the "Copy" buttons next to email addresses?. I.e. button is temporarily replaced by message that reads:
> 
> > < checkmark icon > Copied to clipboard

Should be fixed now

**Comment by Katherine-Osos** [comment]
@CocoByte A few more bits of feedback:

- Only 1 additional contact is getting copied. e.g. try copying info from [he-hundred.gov](http://he-hundred.gov/) on backup sandbox, which has 2 contacts listed.
- Vertically center the text in the “copy request summary” button
- Text edits to ensure wording matches the admin:
  - “Existing Website(s)” should read "Current Websites"
  - “Alternate Domain(s)” should read “Alternative Domains” 
  - “Additional Contact(s)” should read “Other Employees” 

For the text edits - sorry I didn't catch those sooner. Just wording updates, but I didn't realize the ticket didn't reflect the wording we use in the admin. I'll update the ticket to match my edits ...

**Review by Katherine-Osos — APPROVED** [review]
LGTM!

**Review by therealslimhsiehdy — COMMENTED** [review]
So I just realised that because I opened the PR I can't approve this PR 🤣 

That being said -- tested the PR on sandbox and looked through the code and it LGTM. You have my approval, but I'd caveat to ask that Rachid or Zander or anyone who is more experienced with JS look through for styling patterns in case I missed something and give their approval as well!

~(Also this wasn't caused by your PR, but with the profile flag on even if my profile is complete when I fill out a domain request, the submitter for the Domain Request stays empty -- once I added a submitter it copys to the clipboard correctly)~ ETA: Figured out the issue and how it'll be resolved, pls ignore!

```
Recommendation:
Organization Type: Federal
Requested Domain: gov-domain-i-want.gov
Current Websites: http://Currentwebsite00.com, http://currentwebsite01.com
Rationale:
Alternative Domains: alt-domain-0.gov, alt-domain-1.gov
Submitter: Rebecca-Analyst Hsieh-Analyst, Cat, +12063931007
Senior Official: SOFirstName SOLastName, SOTitle, So@so.com
Other Employees:
[OtherEmployeeFirst OtherEmployeeLast](https://getgov-rh.app.cloud.gov/admin/registrar/domainrequest/251/change/?_changelist_filters=q%3Di-want#), OtherEmployeeTitle, OtherEmployee@OE.com, +12063931007
```

File: src/registrar/assets/sass/_theme/_admin.scss
Comment (CocoByte) [reply src/registrar/assets/sass/_theme/_admin.scss:374]: In general adding stray pixel adjustments is indicative of poor css architecture.  After cleaning up alignment and class consistency, the extra 2px was causing some issues so it was removed.

File: src/registrar/assets/js/get-gov-admin.js
Comment (CocoByte) [reply src/registrar/assets/js/get-gov-admin.js:704]: The ticket specified a couple blank fields.  This is one of them.

File: src/registrar/assets/js/get-gov-admin.js
Comment (CocoByte) [reply src/registrar/assets/js/get-gov-admin.js:700]: Same reason as before :)

File: src/registrar/assets/js/get-gov-admin.js
Comment (CocoByte) [reply src/registrar/assets/js/get-gov-admin.js:701]: In general I have always considered it best practice not to mix plain text formatting with html formatting when creating rich text.  </br> keeps things consistent and avoids any risk of the text getting interpreted strangely (even though </br> and \n *should* produce the same result)

**Pull request merged**
